go/proxyd/methods.go

```diff
@@ -193,3 +326,129 @@ func validBlockInput(input string) bool {
 	_, err := decodeBlockInput(input)
 	return err == nil
 }
+
+func makeRPCRes(req *RPCReq, result interface{}) *RPCRes {
+	return &RPCRes{
+		JSONRPC: JSONRPCVersion,
+		ID:      req.ID,
+		Result:  result,
+	}
+}
+
+func copyResError(err *RPCErr) *RPCErr {
+	if err == nil {
+		return nil
+	}
+	return &RPCErr{
+		Code:          err.Code,
+		Message:       err.Message,
+		HTTPErrorCode: err.HTTPErrorCode,
+	}
+}
+
+func copyRes(res *RPCRes) *RPCRes {
+	return &RPCRes{
+		JSONRPC: res.JSONRPC,
+		Result:  res.Result,
+		Error:   copyResError(res.Error),
+		ID:      res.ID,
+	}
+}
+
+type CachedRPC struct {
+	BlockNum   uint64  `json:"blockNum"`
+	Res        *RPCRes `json:"res"`
+	Expiration int64   `json:"expiration"` // in millis since epoch
+}
+
+func (c *CachedRPC) Encode() []byte {
+	return mustMarshalJSON(c)
+}
+
+func (c *CachedRPC) Decode(b []byte) error {
+	return json.Unmarshal(b, c)
+}
+
+func (c *CachedRPC) ExpirationTime() time.Time {
+	return time.Unix(0, c.Expiration*int64(time.Millisecond))
+}
+
+func getImmutableRPCResponse(ctx context.Context, cache Cache, key string, req *RPCReq) (*RPCRes, error) {
+	encodedVal, err := cache.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if encodedVal == "" {
+		return nil, nil
+	}
+	val, err := snappy.Decode(nil, []byte(encodedVal))
+	if err != nil {
+		return nil, err
+	}
+
+	res := new(RPCRes)
+	if err := json.Unmarshal(val, res); err != nil {
+		return nil, err
+	}
+	res.ID = req.ID
+	return res, nil
+}
+
+func putImmutableRPCResponse(ctx context.Context, cache Cache, key string, req *RPCReq, res *RPCRes) error {
+	if key == "" {
+		return nil
+	}
+	val := mustMarshalJSON(res)
+	encodedVal := snappy.Encode(nil, val)
+	return cache.Put(ctx, key, string(encodedVal))
+}
+
+func getBlockDependentCachedRPCResponse(ctx context.Context, cache Cache, getLatestBlockNumFn GetLatestBlockNumFn, key string, req *RPCReq) (*RPCRes, error) {
+	encodedVal, err := cache.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if encodedVal == "" {
+		return nil, nil
+	}
+	val, err := snappy.Decode(nil, []byte(encodedVal))
```

Comment:
I could see a world where there is a `CacheWithCompression` or something and all the calls to `snappy` are wrapped in `Get`/`Put`

Reply:
a fine idea. will do